RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc

```diff
@@ -719,57 +1173,114 @@ TauVars AntiElectronIDMVA6<TauType, ElectronType>::getTauVarsTypeSpecific(const
   // Check type of candidates building tau to avoid dynamic casts further
   bool isFromPFCands =
       (theTau.leadCand().isNonnull() && dynamic_cast<const reco::PFCandidate*>(theTau.leadCand().get()) != nullptr);
-  for (const auto& candidate : theTau.signalCands()) {
-    float etaAtECalEntrance = candidate->eta();
-    float phiAtECalEntrance = candidate->phi();
-    const reco::Track* track = nullptr;
-    if (isFromPFCands) {
-      const reco::PFCandidate* pfCandidate = static_cast<const reco::PFCandidate*>(candidate.get());
-      etaAtECalEntrance = pfCandidate->positionAtECALEntrance().eta();
-      if (!usePhiAtEcalEntranceExtrapolation_) {
-        phiAtECalEntrance = pfCandidate->positionAtECALEntrance().phi();
+  if (!isPhase2_) {
+    for (const auto& candidate : theTau.signalCands()) {
+      float etaAtECalEntrance = candidate->eta();
+      float phiAtECalEntrance = candidate->phi();
+      const reco::Track* track = nullptr;
+      if (isFromPFCands) {
+        const reco::PFCandidate* pfCandidate = static_cast<const reco::PFCandidate*>(candidate.get());
+        etaAtECalEntrance = pfCandidate->positionAtECALEntrance().eta();
+        if (!usePhiAtEcalEntranceExtrapolation_) {
+          phiAtECalEntrance = pfCandidate->positionAtECALEntrance().phi();
+        } else {
+          bool success = false;
+          reco::Candidate::Point posAtECal = positionAtECalEntrance_(candidate.get(), success);
+          if (success) {
+            phiAtECalEntrance = posAtECal.phi();
+          }
+        }
+        if (pfCandidate->trackRef().isNonnull())
+          track = pfCandidate->trackRef().get();
+        else if (pfCandidate->muonRef().isNonnull() && pfCandidate->muonRef()->innerTrack().isNonnull())
+          track = pfCandidate->muonRef()->innerTrack().get();
+        else if (pfCandidate->muonRef().isNonnull() && pfCandidate->muonRef()->globalTrack().isNonnull())
+          track = pfCandidate->muonRef()->globalTrack().get();
+        else if (pfCandidate->muonRef().isNonnull() && pfCandidate->muonRef()->outerTrack().isNonnull())
+          track = pfCandidate->muonRef()->outerTrack().get();
+        else if (pfCandidate->gsfTrackRef().isNonnull())
+          track = pfCandidate->gsfTrackRef().get();
```

Comment:
given a repeat below, please refactor to use a common function

Reply:
OK